### ISSUE TITLE
Performance regression for FP16 Relu kernel from TF2.6 to TF2.7

### ISSUE THREAD

**Issue by wenscarl** [keras-team/keras#15857]
<html>
<body>
<p dir="auto" style="box-sizing: border-box; margin-top: 0px !important; margin-bottom: 16px; color: rgb(36, 41, 47); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Migrating from TF2.6 to TF2.7 FP16 Relu kernel has changed from EigenMetaKernel to MLIR.<br style="box-sizing: border-box;">Tested on Volta and Ampere, the memory throughput gets halved and compute throughput is also reduced. Collected by nsight compute.</p>

Throughput % | compute | memory
-- | -- | --
EigenMetaKernel | 35.79 | 84.20
Relu_GPU_DT_HALF_DT_HALF_kernel | 28.17 | 45.02

<p dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 0px !important; color: rgb(36, 41, 47); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">A roughly 2x slowdown of the forward pass is observed. The better bandwidth utilization of EigenMetaKernel is contributed from wider vectors for GPU-half implementation for libeigen.

A repo code is,
import numpy as np
import tensorflow as tf

from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras import mixed_precision

policy = mixed_precision.Policy('mixed_float16')
mixed_precision.set_global_policy(policy)

inputs = keras.Input(shape=(6400,), name='digits')
num_units = 8192

dense1 = layers.Dense(num_units, activation='relu', name='dense_1')
x = dense1(inputs)
dense2 = layers.Dense(num_units, activation='relu', name='dense_2')
x = dense2(x)

x = layers.Dense(1000, name='dense_logits')(x)
outputs = layers.Activation('relu', dtype='float32', name='predictions')(x)

model = keras.Model(inputs=inputs, outputs=outputs)
model.compile(loss='sparse_categorical_crossentropy',
              optimizer=keras.optimizers.RMSprop(),
              metrics=['accuracy'])

x_train = np.random.normal(size=(60000, 6400)).astype('float32') / 255
y_train = np.random.randint(10, size=60000)

history = model.fit(x_train, y_train, batch_size=8192, epochs=5, validation_split=0.2)

On TF2.7, nsight-compute profile is collected by
`ncu -k Relu_GPU_DT_HALF_DT_HALF_kernel --target-processes all -f -o repo_TF27 python repo.py`
On TF2.6, `ncu -k EigenMetaKernel --target-processes all -f -o repo_TF26 python repo.py..` For this example, the mem throughput is about 88%(TF2.7) VS 68%(TF2.6). The demangled name for EigenMetaKernel is
"void Eigen::internal::EigenMetaKernel<Eigen::TensorEvaluator<const Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<Eigen::half, (int)1, (int)1, long>, (int)16, Eigen::MakePointer>, const Eigen::TensorCwiseBinaryOp<Eigen::internal::scalar_max_op<const Eigen::half, const Eigen::half, (int)1>, const Eigen::TensorMap<Eigen::Tensor<const Eigen::half, (int)1, (int)1, long>, (int)16, Eigen::MakePointer>, const Eigen::TensorCwiseNullaryOp<Eigen::internal::scalar_constant_op<const Eigen::half>, const Eigen::TensorMap<Eigen::Tensor<const Eigen::half, (int)1, (int)1, long>, (int)16, Eigen::MakePointer>>>>, Eigen::GpuDevice>, long>(T1, T2

</p>
</body>
</html>

**Comment by qlzh727** [comment]
Triage notes: Assign to Reed for FP16 related quesitons.

**Comment by reedwm** [comment]
I cannot reproduce on a Titan V with CUDA 11.2.2 and cuDNN 8.1.1. With both TF 2.6 and 2.7, the training throughput is roughly 180 ms/s, and looking in the Tensorboard profile, relu takes approximately the same amount of time, although I do see the new MLIR kernel in TF 2.7. Note in your example I had to add the line `x_train = np.random.normal(size=(60000, 6400))` to define `x_train`.

Can you clarify exactly what GPUs and CUDA versions you are using? Also, can you try running the following dedicated relu benchmark to see if TF 2.7 is slower?

```python
import tensorflow as tf
import time

x = tf.random.normal((8192, 8192), dtype=tf.float16)
p = tf.constant(0.)
tf.nn.relu(x)  # Warmup
start = time.time()
for _ in range(1000):
  tf.nn.relu(x)
(p + 1.).numpy()  # Synchronize GPU
end = time.time()

print(f'Time taken: {end - start}')
```

**Comment by wenscarl** [comment]
I am running on V100x8 with cuda-11.5 and cudnn 8.3.1. For the relu benchmark above, I still see difference. In TF2.7, time taken is 0.43 compared to 0.34 in TF2.6. nsys profiler tells that the eignMetaKernel takes average 344us (TF2.6) vs 446us(TF2.7).

**Comment by reedwm** [comment]
The benchmark prints the time, so no need to get the time from a profiler. In any case, I do not have a new enough driver for cuda 11.5, but i tried on cuda 11.4. On the relu benchmark on a Titan V, I got 0.465 seconds and 0.487 seconds on TF 2.6 and TF 2.7 respectively, meaning there is a 5% regression on TF 2.7. On a Titan RTX, I got 0.521 seconds and 0.599 seconds respectively, which is a 15% regression on TF 2.7.

I could only reproduce a 15% regression, but this is still significant. @sherhut, can you take a look? I cannot assign the GitHub issue to you, probably because you not part of the Keras GitHub team.

**Comment by tlemo** [comment]
@wenscarl, can you please update the repro script to fix the missing `x_train` definition? (as pointed out by @reedwm as well)

**Comment by wenscarl** [comment]
@tlemo It's updated.

**Comment by chsigg** [comment]
This should be fixed now with the these two changes:

https://reviews.llvm.org/D117219 [LLVM] vectorize vector loads
[tensorflow/b826391](https://github.com/tensorflow/tensorflow/commit/b8263919112cb63e4389dbd0c7b39e27a30cc17d): unroll relu kernel

The following changes improve the generated code further, but don't change kernel performance because it's already memory bound:

[tensorflow/271845a](https://github.com/tensorflow/tensorflow/commit/271845ad0af96b8fb4de6ade92acb0961fd31785): generate sm_80 code
https://reviews.llvm.org/D117010: [MLIR] Mark arith.minf, arith.maxf as commutative
https://reviews.llvm.org/D117011: [MLIR] Remove NaN from arith.maxf expansion
https://reviews.llvm.org/D117101: [MLIR] Fold cmpf with NaN
https://reviews.llvm.org/D117204: [NVPTX] Lower to fmax.NaN for sm_80+.
https://reviews.llvm.org/D118126: [LLVM] instcombine into fmaximum (pending)

**Comment by wenscarl** [comment]
Verified on tf-nightly, the memory throughput for `Relu_GPU_DT_HALF_DT_HALF_kernel` increases to ~88%. Closing. Thanks! @chsigg
